dist/function.py

```diff
@@ -189,8 +189,161 @@ def function_declaration():
             tfile.close()
             compare_srcfile(tmp_file, name)
 
+# A function definition.
+class FunctionDefinition:
+    def __init__(self, name, module, source):
+        self.name = name
+        self.module = module
+        self.source = source
+        self.used_outside_of_file = False
+        self.used_outside_of_module = False
+
+# Complain if a "__wti" function is used outside of its intended scope, or if a "__wt" function is
+# used only within its module.
+def function_scoping():
+    skip_re = re.compile(r'DO NOT EDIT: automatically built')
+    func_def_re = re.compile(r'\n\w[\w \*]+\n(\w+)', re.DOTALL)
+    func_use_re = re.compile(r'[^\n](\w+)', re.DOTALL)
+    failed = False
+
+    # Infer the module name from the file path.
+    def infer_module(path):
+        parts = path.split('/')
+        if len(parts) < 3 or parts[0] != '..':
+            print(f'{path}: Unexpected path')
+            sys.exit(1)
+
+        # If the path is in src, the module is the subdirectory under src; otherwise it is the
+        # top-level directory, such as "test."
+        if parts[1] == 'src':
+            module = parts[2]
+        else:
+            module = parts[1]
+
+        # Make "os_posix" and "os_win" one module because they declare the same "__wt" functions.
+        if module == 'os_posix' or module == 'os_win':
+            return 'os_posix/os_win'
+        return module
+
+    # Find all "__wt" and "__wti" function definitions.
+    fn_defs = dict()
+    for source_file in source_files():
+        if not source_file.startswith('../src/'):
+            continue
+        module = infer_module(source_file)
+
+        # Skip auto-generated files.
+        file_contents = open(source_file, 'r').read()
+        if skip_re.search(file_contents):
+            continue
+
+        # Now actually find all definitions.
+        for m in func_def_re.finditer(file_contents):
+            fn_name = m.group(1)
+            if not fn_name.startswith('__wt_') and not fn_name.startswith('__wti_'):
```

Comment:
These lines could be omitted since only functions that are start with __wt or __wti are in fn_defs. Would that be faster? How does two starswith() compare with one in dict()? If you delete these lines add a comment that functions that are not __wt and __wti are filtered out by in fnd_defs.

Reply:
I was wrong. These lines are needed for line 249. They could be changed to the below and left here of moved just before lien 249.

   global_scope = fn_name.startswith('__wt_')
   if not global_scope and not fn_name.startswith('__wti_'):
        continue

Then change line 249
     fn_defs[fn_name] = FunctionDefinition(fn_name, module, source_file, global_scope)

Instead of checking fn_name.startswith('__wt__') check fn_defs[fn_name].global_scope.
Instead of checking fn_name.startswith('__wti__') check not fn_defs[fn_name].global_scope.